src/apps/monero/protocol/signing/step_01_init_transaction.py

```diff
@@ -0,0 +1,242 @@
+"""
+Initializes a new transaction.
+"""
+
+import gc
+
+from apps.monero.controller import misc
+from apps.monero.layout import confirms
+from apps.monero.protocol.signing.state import State
+from apps.monero.xmr import common, crypto, monero
+
+
+async def init_transaction(state: State, address_n, network_type, tsx_data):
+    from apps.monero.xmr.sub.addr import classify_subaddresses
+    from apps.monero.protocol import hmac_encryption_keys
+
+    state.creds = await misc.monero_get_creds(state.ctx, address_n, network_type)
+
+    state.fee = state.fee if state.fee > 0 else 0
+
+    state.tx_priv = crypto.random_scalar()
+    state.tx_pub = crypto.scalarmult_base(state.tx_priv)
+
+    state.mem_trace(1)
+
+    # Ask for confirmation
+    await confirms.confirm_transaction(state.ctx, tsx_data, state.creds)
+    gc.collect()
+    state.mem_trace(3)
+
+    # Basic transaction parameters
+    state.input_count = tsx_data.num_inputs
+    state.output_count = len(tsx_data.outputs)
+    state.output_change = tsx_data.change_dts
+    state.mixin = tsx_data.mixin
+    state.fee = tsx_data.fee
+    state.account_idx = tsx_data.account
+    state.multi_sig = tsx_data.is_multisig
+    check_change(state, tsx_data.outputs)
+    state.exp_tx_prefix_hash = tsx_data.exp_tx_prefix_hash
+
+    # Rsig data
+    state.rsig_type = tsx_data.rsig_data.rsig_type
+    state.rsig_grp = tsx_data.rsig_data.grouping
+    state.rsig_offload = state.rsig_type > 0 and state.output_count > 2
+    state.use_bulletproof = state.rsig_type > 0
+    state.use_simple_rct = state.input_count > 1 or state.rsig_type != 0
+
+    # Provided tx key, used mostly in multisig.
+    if len(tsx_data.use_tx_keys) > 0:
+        for ckey in tsx_data.use_tx_keys:
+            crypto.check_sc(crypto.decodeint(ckey))
+
+        state.tx_priv = crypto.decodeint(tsx_data.use_tx_keys[0])
+        state.tx_pub = crypto.scalarmult_base(state.tx_priv)
+        state.additional_tx_private_keys = [
+            crypto.decodeint(x) for x in tsx_data.use_tx_keys[1:]
+        ]
+
+    # Additional keys w.r.t. subaddress destinations
+    class_res = classify_subaddresses(tsx_data.outputs, state.change_address())
+    num_stdaddresses, num_subaddresses, single_dest_subaddress = class_res
+
+    # if this is a single-destination transfer to a subaddress, we set the tx pubkey to R=r*D
+    if num_stdaddresses == 0 and num_subaddresses == 1:
+        state.tx_pub = crypto.scalarmult(
+            crypto.decodepoint(single_dest_subaddress.spend_public_key), state.tx_priv
+        )
+
+    state.need_additional_txkeys = num_subaddresses > 0 and (
+        num_stdaddresses > 0 or num_subaddresses > 1
+    )
+    state.mem_trace(4, True)
+
+    # Extra processing, payment id
+    state.tx.version = 2
```

Comment:
Why 2? Current Monero transaction format?

Reply:
Yes, current transaction version in the Monero.